File: crates/sinkron-common/src/protocol.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::error::SinkronError;

/*
Client connects to server via webscokets. All messages must be prefixed
with an integer specifying a channel. It allows to synchronize multiple
collections at the same time using one connection.

Channel "0" is reserved for system messages.

The first message in the newly opened channel from the client must always be
"sync_start" message.

In case of success server replies with document updates messages required
to achieve synchronized state ("doc", "update" and "delete"), followed by
"sync_complete" message.

In case of error server sends "sync_error" message, after which the channel is
closed.

After successful synchronization client can start sending its own updates and
recieve realtime updates performed by other clients.

Invalid messages, e.g. messages to a closed channel, are simply ignored.
*/

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct HeartbeatMessage {
    pub i: i32,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct SyncStartMessage {
    pub col: String,
    pub colrev: i64,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct SyncStopMessage {
    pub col: String,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct SyncErrorMessage {
    pub col: String,
    #[serde(flatten)]
    pub error: SinkronError,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct SyncCompleteMessage {
    pub col: String,
    pub colrev: i64,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct GetMessage {
    pub col: String,
    pub id: Uuid,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct GetErrorMessage {
    pub col: String,
    pub id: Uuid,
    #[serde(flatten)]
    pub error: SinkronError,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct ClientCreateMessage {
    pub col: String,
    pub id: Uuid,
    pub content: String,
    pub files: Vec<Uuid>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct FilesUpdate {
    pub add: Vec<Uuid>,
    pub delete: Vec<Uuid>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClientUpdateMessage {
    pub col: String,
    pub id: Uuid,
    pub content_update: Option<String>,
    pub files_update: Option<FilesUpdate>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct ClientDeleteMessage {
    pub col: String,
    pub id: Uuid,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DocMessage {
    pub id: Uuid,
    pub col: String,
    pub colrev: i64,
    pub content: String,
    pub files: Vec<Uuid>,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServerUpdateMessage {
    pub col: String,
    pub id: Uuid,
    pub colrev: i64,
    pub content_update: Option<String>,
    pub files: Vec<Uuid>,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct ServerDeleteMessage {
    pub col: String,
    pub id: Uuid,
    pub colrev: i64,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct ChangeErrorMessage {
    pub col: String,
    pub id: Uuid,
    #[serde(flatten)]
    pub error: SinkronError,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum ClientMessage {
    #[serde(rename = "h")]
    Heartbeat(HeartbeatMessage),

    #[serde(rename = "sync_start")]
    SyncStart(SyncStartMessage),

    #[serde(rename = "sync_stop")]
    SyncStop(SyncStopMessage),

    #[serde(rename = "get")]
    Get(GetMessage),

    #[serde(rename = "create")]
    Create(ClientCreateMessage),

    #[serde(rename = "update")]
    Update(ClientUpdateMessage),

    #[serde(rename = "delete")]
    Delete(ClientDeleteMessage),
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum ServerMessage {
    #[serde(rename = "connection_error")]
    ConnectionError(SinkronError),

    #[serde(rename = "h")]
    Heartbeat(HeartbeatMessage),

    #[serde(rename = "sync_complete")]
    SyncComplete(SyncCompleteMessage),

    #[serde(rename = "sync_error")]
    SyncError(SyncErrorMessage),

    #[serde(rename = "get_error")]
    GetError(GetErrorMessage),

    #[serde(rename = "doc")]
    Doc(DocMessage),

    #[serde(rename = "update")]
    Update(ServerUpdateMessage),

    #[serde(rename = "delete")]
    Delete(ServerDeleteMessage),

    #[serde(rename = "change_error")]
    ChangeError(ChangeErrorMessage),
}

pub fn parse_channel_prefix(input: &str) -> Option<(i32, &str)> {
    // String must start with a number, followed by ":" symbol
    let mut len = 0;
    for c in input.chars() {
        if char::is_numeric(c) {
            len += 1;
        } else {
            break;
        }
    }
    if len == 0 {
        return None;
    }
    if input.chars().nth(len) != Some(':') {
        return None;
    }
    let Ok(prefix) = str::parse::<i32>(&input[0..len]) else {
        return None;
    };
    return Some((prefix, &input[len + 1..]));
}

pub fn serialize_with_channel_prefix<T: Serialize>(
    channel: i32,
    msg: &T,
) -> Option<String> {
    let mut res = channel.to_string() + ":";
    match serde_json::to_writer(unsafe { res.as_mut_vec() }, &msg) {
        Ok(_) => Some(res),
        Err(_) => None,
    }
}

File: packages/sinkron-client/src/protocol.ts
// General errors
export type BadRequestError = {
    code: "bad_request"
    message: string
}

export type AuthFailedError = {
    code: "auth_failed"
    message: string
}

export type NotFoundError = {
    code: "not_found"
    message: string
}

export type ForbiddenError = {
    code: "forbidden"
    message: string
}

export type InternalServerError = {
    code: "internal_server_error"
    message: string
}

export type UnprocessableContentError = {
    code: "unprocessable_content"
    message: string
}

// Collection & document errors
export type InvalidColrevError = {
    code: "invalid_colrev"
    // col_id
}

export type DuplicateDocumentIdError = {
    code: "duplicate_document"
    // col_id, doc_id
}

export type DocumentAlreadyDeletedError = {
    code: "document_already_deleted"
    // col_id, doc_id
}

export type InsufficientStorageError = {
    code: "insufficient_storage"
    // col_id, remaining_storage
}

export type ContentTooLargeError = {
    code: "content_too_large"
    // col_id
}

// File errors
export type FileNotFoundError = {
    code: "file_not_found"
    file_id: string
}

export type FileMissingChunksError = {
    code: "file_missing_chunks"
    file_id: string
}

export type FileTooLargeError = {
    code: "file_too_large"
    file_id: string
}

export type SinkronError =
    | BadRequestError
    | AuthFailedError
    | NotFoundError
    | ForbiddenError
    | InternalServerError
    | UnprocessableContentError
    | InvalidColrevError
    | DuplicateDocumentIdError
    | DocumentAlreadyDeletedError
    | InsufficientStorageError
    | ContentTooLargeError
    | FileNotFoundError
    | FileMissingChunksError
    | FileTooLargeError

export type ConnectionErrorMessage = {
    kind: "connection_error"
    i: number
}

export type HeartbeatMessage = {
    kind: "h"
    i: number
}

export type SyncStartMessage = {
    kind: "sync_start"
    col: string
    colrev: number
}

export type SyncStopMessage = {
    kind: "sync_stop"
    col: string
}

export type SyncErrorMessage = SinkronError & {
    kind: "sync_error"
    col: string
}

export type SyncCompleteMessage = {
    kind: "sync_complete"
    col: string
    colrev: number
}

export type GetMessage = {
    kind: "get"
    col: string
    id: string
}

export type GetErrorMessage = SinkronError & {
    kind: "get_error"
    col: string
    id: string
}

export type ClientCreateMessage = {
    kind: "create"
    col: string
    id: string
    content: string
    files: string[]
}

export type FilesUpdate = {
    add: string[]
    delete: string[]
}

export type ClientUpdateMessage = {
    kind: "update"
    col: string
    id: string
    content_update: string | null
    files_update: FilesUpdate | null
}

export type ClientDeleteMessage = {
    kind: "delete"
    col: string
    id: string
}

export type DocMessage = {
    kind: "dec"
    id: string
    col: string
    colrev: number
    content: string
    files: File[]
    created_at: string // iso8601
    updated_at: string // iso8601
}

export type ServerUpdateMessage = {
    kind: "update"
    col: string
    id: string
    colrev: number
    content_update: string | null
    files: File[]
    created_at: string // iso8601
    updated_at: string // iso8601
}

export type ServerDeleteMessage = {
    kind: "delete"
    col: string
    id: string
    colrev: number
}

export type ChangeErrorMessage = SinkronError & {
    kind: "change_error"
    col: string
    id: string
}

export type ClientMessage =
    | HeartbeatMessage
    | SyncStartMessage
    | SyncStopMessage
    | GetMessage
    | ClientCreateMessage
    | ClientUpdateMessage
    | ClientDeleteMessage

export type ServerMessage =
    | ConnectionErrorMessage
    | HeartbeatMessage
    | SyncCompleteMessage
    | SyncErrorMessage
    | GetErrorMessage
    | DocMessage
    | ServerUpdateMessage
    | ServerDeleteMessage
    | ChangeErrorMessage
